rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala

```diff
@@ -0,0 +1,591 @@
+            val Channel(Quote(parentKeyPar))                   = parentKey
+            val Channel(Quote(parentDataPar))                  = parentData
+            val Some(Expr(EMapBody(parMap)))                   = parentDataPar.singleExpr()
+            val Some(parentEntry)                              = parMap.ps.get(parentKeyPar)
+            val Some(Expr(ETupleBody(ETuple(parentPs, _, _)))) = parentEntry.singleExpr()
+            if (parentPs.length != 3)
+              localFail()
+            else {
+              val Some(Expr(GByteArray(parentAdditional))) = parentPs(1).singleExpr
+              val Some(Expr(GByteArray(lastKeyStr)))       = lastKey.singleExpr()
+              val Some(Expr(ETupleBody(ETuple(ps, _, _)))) = lastEntry.singleExpr()
+              if (ps.length != 3)
+                localFail()
+              else {
+                val Some(Expr(GByteArray(edgeAdditional))) = ps(1).singleExpr()
+                val mergeStream                            = ByteString.newOutput()
+                parentAdditional.writeTo(mergeStream)
+                lastKeyStr.writeTo(mergeStream)
+                edgeAdditional.writeTo(mergeStream)
+                val mergedEdge        = parByteArray(mergeStream.toByteString())
+                val updatedTuple: Par = ETuple(Seq(ps(0), mergedEdge, ps(2)))
+                val updatedMap: Par = ParMap(
+                  SortedParMap(parMap.ps + (parentKeyPar -> updatedTuple)))
+                replace(Quote(updatedMap), parentReplace, parentRand)
+              }
+            }
+          }
+
+          val Channel(Quote(dataPar))                  = data
+          val Some(Expr(EMapBody(parMap)))             = dataPar.singleExpr
+          val Some(value)                              = parMap.ps.get(parByteArray(head))
+          val Some(Expr(ETupleBody(ETuple(ps, _, _)))) = value.singleExpr()
+          if (ps.length != 3)
+            localFail()
+          else {
+            // The second tuple field should be a bytearray in both cases.
+            val Some(Expr(GByteArray(edgeAdditional))) = ps(1).singleExpr()
+            ps(0).singleExpr() match {
+              case Some(Expr(GInt(0))) =>
+                if (tail == edgeAdditional) {
+                  if (parMap.ps.size > 2) {
+                    val updatedMap: Par = ParMap(SortedParMap(parMap.ps - parByteArray(head)))
+                    for {
+                      _ <- replace(Quote(updatedMap), replaceChan, dataRand)
+                      _ <- replace(parentData, parentReplace, parentRand)
+                      _ <- succeed(ret, ps(2), callRand)
+                    } yield ()
+                  } else if (parMap.ps.size != 2) {
+                    localFail()
+                  } else {
+                    val it = (parMap.ps - parByteArray(head)).iterator
+                    for {
+                      _ <- Function.tupled(mergeWithParent(_, _))(it.next)
```

Comment:
Why complicate it with an `Iterator` ? 
```scala
val it = parMap.ps - parByteArray(head)
  for {
    _ <- Function.tupled(mergeWithParent(_, _))(it.head)
```
works just the same

Reply:
Done.